File: app/src/main/java/org/tasks/tasklist/DiffCallback.java
package org.tasks.tasklist;

import androidx.recyclerview.widget.DiffUtil;
import com.todoroo.astrid.adapter.TaskAdapter;
import java.util.List;
import org.tasks.data.TaskContainer;

class DiffCallback extends DiffUtil.Callback {

  private final List<TaskContainer> oldList;
  private final List<TaskContainer> newList;
  @Deprecated private final TaskAdapter adapter;

  DiffCallback(List<TaskContainer> oldList, List<TaskContainer> newList, TaskAdapter adapter) {
    this.oldList = oldList;
    this.newList = newList;
    this.adapter = adapter;
  }

  @Override
  public int getOldListSize() {
    return oldList.size();
  }

  @Override
  public int getNewListSize() {
    return newList.size();
  }

  @Override
  public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
    return oldList.get(oldItemPosition).getId() == newList.get(newItemPosition).getId();
  }

  @Override
  public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
    TaskContainer oldItem = oldList.get(oldItemPosition);
    TaskContainer newItem = newList.get(newItemPosition);
    return oldItem.equals(newItem) && oldItem.getIndent() == adapter.getIndent(newItem);
  }
}

File: app/src/main/java/org/tasks/tasklist/DiffCallback.kt
package org.tasks.tasklist

import androidx.recyclerview.widget.DiffUtil
import com.todoroo.astrid.adapter.TaskAdapter
import com.todoroo.astrid.core.SortHelper.SORT_DUE
import com.todoroo.astrid.core.SortHelper.SORT_START
import timber.log.Timber

internal class DiffCallback(
    private val old: SectionedDataSource,
    private val new: SectionedDataSource,
    @Deprecated("") private val adapter: TaskAdapter
) : DiffUtil.Callback() {

    private val refreshDates = when (old.groupMode) {
        SORT_DUE -> new.groupMode == SORT_START
        SORT_START -> new.groupMode == SORT_DUE
        else -> false
    }

    override fun getOldListSize() = old.size

    override fun getNewListSize() = new.size

    override fun areItemsTheSame(oldPosition: Int, newPosition: Int): Boolean {
        val wasHeader = old.isHeader(oldPosition)
        val isHeader = new.isHeader(newPosition)
        if (wasHeader != isHeader) {
            return false
        }
        return try {
            if (isHeader) {
                old.groupMode == new.groupMode && old.getHeaderValue(oldPosition) == new.getHeaderValue(newPosition)
            } else {
                old.getItem(oldPosition).id == new.getItem(newPosition).id
            }
        } catch (e: IndexOutOfBoundsException) {
            Timber.w(e, "areItemsTheSame: old=$oldPosition/${old.size} new=$newPosition/${new.size}")
            false
        }
    }

    override fun areContentsTheSame(oldPosition: Int, newPosition: Int): Boolean {
        return try {
            if (new.isHeader(newPosition)) {
                old.getSection(oldPosition).collapsed == new.getSection(newPosition).collapsed
            } else {
                val oldItem = old.getItem(oldPosition)
                val newItem = new.getItem(newPosition)
                !refreshDates && oldItem == newItem && oldItem.indent == adapter.getIndent(newItem)
            }
        } catch (e: IndexOutOfBoundsException) {
            Timber.w(e, "areContentsTheSame: old=$oldPosition/${old.size} new=$newPosition/${new.size}")
            false
        }
    }
}